File: src/main/java/editor/database/card/TransformCard.java
package editor.database.card;

import java.util.List;

import editor.database.attributes.ManaCost;
import editor.util.Lazy;

/**
 * This class represents a Card with two faces:  One on the front, and one on the back.
 * During a game, it will transform from one to the other.
 *
 * @author Alec Roelke
 */
public class TransformCard extends MultiCard
{
    /**
     * Card representing the front face.
     */
    private final Card front;
    /**
     * Tuple of this DoubleFacedCard's faces' mana costs.
     */
    private Lazy<List<ManaCost>> manaCost;

    /**
     * Create a new TransformCard with the given Cards as faces.  Their layouts should
     * say that they are double-faced cards.
     *
     * @param f card representing the front face
     * @param b card representing the back face
     */
    public TransformCard(Card f, Card b)
    {
        super(CardLayout.TRANSFORM, f, b);
        front = f;
        if (front.layout() != CardLayout.TRANSFORM || b.layout() != CardLayout.TRANSFORM)
            throw new IllegalArgumentException("can't join non-double-faced cards into double-faced cards");

        manaCost = new Lazy<>(() -> List.of(front.manaCost().get(0), new ManaCost()));
    }

    /**
     * {@inheritDoc}
     * A transform card's mana value is that of its front face.
     */
    @Override
    public double cmc()
    {
        return front.cmc();
    }

    /**
     * {@inheritDoc}
     * Only the front face has a mana cost.
     */
    @Override
    public List<ManaCost> manaCost()
    {
        return manaCost.get();
    }
}

File: src/main/scala/editor/database/card/TransformCard.scala
package editor.database.card

import editor.database.attributes.ManaCost

import CardLayout.TRANSFORM

/**
 * A [[Card]] with two faces that can transform back and forth between them during a game. Its mana value
 * is that of its front face, but it has two mana costs: that of its front face, and an empty cost for
 * the back face.
 * 
 * @constructor create a new transforming card
 * @param front front of the card (what it is when off the battlefield)
 * @param back back of the card (what it can transform into)
 * 
 * @author Alec Roelke
 */
@throws[IllegalArgumentException]("if either face is not a transform card")
class TransformCard(front: Card, back: Card) extends MultiCard(TRANSFORM, IndexedSeq(front, back)) {
  if (front.layout != TRANSFORM || back.layout != TRANSFORM)
    throw IllegalArgumentException("can't join non-transforming cards into transforming cards")

  override def manaValue = front.manaValue
  override def minManaValue = manaValue
  override def maxManaValue = manaValue
  override def avgManaValue = manaValue
  override lazy val manaCost = front.manaCost
  override lazy val imageNames = front.imageNames ++ back.imageNames
}